services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/registry.py

```diff
@@ -46,47 +50,94 @@ async def _is_registry_reachable(registry_settings: RegistrySettings) -> None:
+            raise _RegistryNotReachableError(error_message)
+
+
+async def wait_for_registries_liveness(app: FastAPI) -> None:
+    settings: ApplicationSettings = app.state.settings
 
     await wait_for_service_liveness(
         _is_registry_reachable,
-        service_name="Registry",
-        endpoint=_get_registry_url(registry_settings),
-        registry_settings=registry_settings,
+        service_name="Internal Registry",
+        endpoint=_get_registry_url(settings.DY_DEPLOYMENT_REGISTRY_SETTINGS),
+        registry_settings=settings.DY_DEPLOYMENT_REGISTRY_SETTINGS,
     )
 
+    if settings.DY_DOCKER_HUB_REGISTRY_SETTINGS:
+        await wait_for_service_liveness(
+            _is_dockerhub_reachable,
+            service_name="DockerHub Registry",
+            endpoint="https://hub.docker.com/v2/users/login",
+            registry_settings=settings.DY_DOCKER_HUB_REGISTRY_SETTINGS,
+        )
 
-async def _login_registry(registry_settings: RegistrySettings) -> None:
+
+async def _login_registries(settings: ApplicationSettings) -> None:
     """
     Creates ~/.docker/config.json and adds docker registry credentials
     """
 
-    def create_docker_config_file(registry_settings: RegistrySettings) -> None:
+    def _get_credentials(registry_settings: RegistrySettings) -> str:
         user = registry_settings.REGISTRY_USER
         password = registry_settings.REGISTRY_PW.get_secret_value()
+        return base64.b64encode(f"{user}:{password}".encode()).decode()
+
+    def create_docker_config_file(registries_settings: list[RegistrySettings]) -> None:
         docker_config = {
             "auths": {
-                f"{registry_settings.resolved_registry_url}": {
-                    "auth": base64.b64encode(f"{user}:{password}".encode()).decode(
-                        "utf-8"
-                    )
+                f"{settings.resolved_registry_url}": {
+                    "auth": _get_credentials(settings)
                 }
+                for settings in registries_settings
             }
         }
-        conf_file = Path.home() / ".docker" / "config.json"
-        conf_file.parent.mkdir(exist_ok=True, parents=True)
-        conf_file.write_text(json.dumps(docker_config))
 
-    if registry_settings.REGISTRY_AUTH:
-        await asyncio.get_event_loop().run_in_executor(
-            None, create_docker_config_file, registry_settings
-        )
+        DOCKER_CONFIG_JSON_PATH.parent.mkdir(exist_ok=True, parents=True)
+        DOCKER_CONFIG_JSON_PATH.write_text(json.dumps(docker_config))
+
+    registries_settings: list[RegistrySettings] = []
+    if settings.DY_DEPLOYMENT_REGISTRY_SETTINGS.REGISTRY_AUTH:
+        registries_settings.append(settings.DY_DEPLOYMENT_REGISTRY_SETTINGS)
+    if settings.DY_DOCKER_HUB_REGISTRY_SETTINGS:
+        registries_settings.append(settings.DY_DOCKER_HUB_REGISTRY_SETTINGS)
+
+    await asyncio.get_event_loop().run_in_executor(
+        None, create_docker_config_file, registries_settings
```

Comment:
I have several questions here.

1. Theoretically this approach is to convert **IO bound** synchronous calls into async calls (see explanations and example in [run_in_executor manual](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.run_in_executor)). AFAIK `create_docker_config_file` is a **CPU bound** call
2. what is the difference between `asyncio.get_event_loop().run_in_executor` and simply [`asyncio.to_thread`](
https://docs.python.org/3/library/asyncio-task.html#asyncio.to_thread). At first sight they seem the same to me and the later might be more clear.
3. Is this optimization really necessary? I can understand that if you have a service attending multiple requests this might be key but the sidecare is a one-job service ...

Reply:
``` python
import asyncio

def foo(x):
    # Simulating a blocking I/O-bound task
    import time
    time.sleep(2)
    return x * 2

async def main1():
    result = await asyncio.to_thread(foo, 10)
    print(result)
    
async def main2():
    loop = asyncio.get_event_loop()
    result = await loop.run_in_executor(None, foo, 10)
    print(result)

t0 = time.time()    
asyncio.run(main1())
t1 = time.time()
asyncio.run(main2())
t2 = time.time()

```

Reply:
@pcrespov writing a file is IO bound. why CPU bound?
And using to_thread in an asyncio system I find less clear. We should use the to_async decorator.

Reply:
- I thought that you are using this for pydantic validation, not writing af ile. 
- My question about `to_thread` is more whether it is equivalent or not. AFAIK, in this case, it would be identical and using a higher level API. IMO if the context is clear, I would rather use a high level API